Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,464 @@
+            CinemachineCore.Stage stage, ref CameraState state, float deltaTime)
+        {
+            if (stage == CinemachineCore.Stage.Body)
+            {
+                if (!m_shapeCache.ValidateCache(m_BoundingShape2D, m_MaxOrthoSize, m_confinerBaker, 
+                    state.Lens.Aspect, out bool confinerStateChanged))
+                {
+                    return; // invalid path
+                }
+                
+                float frustumHeight = CalculateHalfFrustumHeight(state, vcam);
+                var extra = GetExtraState<VcamExtraState>(vcam);
+                extra.m_vcamShapeCache.ValidateCache(m_confinerBaker, 
+                    confinerStateChanged, frustumHeight, state.Lens.Orthographic);
+
+
+                Vector3 cameraPosLocal = m_shapeCache.TransformPointToConfinerSpace(state.CorrectedPosition);
+                Vector3 displacement = ConfinePoint(cameraPosLocal, extra.m_vcamShapeCache.m_path);
+                displacement = m_shapeCache.TransformConfinerSpacePointToWorld(displacement);
+
+                // Remember the desired displacement for next frame
+                var prev = extra.m_previousDisplacement;
+                extra.m_previousDisplacement = displacement;
+
+                if (!VirtualCamera.PreviousStateIsValid || deltaTime < 0 || m_Damping <= 0)
+                    extra.m_dampedDisplacement = Vector3.zero;
+                else
+                {
+                    // If a big change from previous frame's desired displacement is detected, 
+                    // assume we are going around a corner and extract that difference for damping
+                    if (Vector2.Angle(prev, displacement) > m_cornerAngleTreshold)
+                        extra.m_dampedDisplacement += displacement - prev;
+
+                    extra.m_dampedDisplacement -= Damper.Damp(extra.m_dampedDisplacement, m_Damping, deltaTime);
+                    displacement -= extra.m_dampedDisplacement;
+                }
+                state.PositionCorrection += displacement;
+            }
+        }
+
+        /// <summary>
+        /// Calculates half frustum height for orthographic or perspective camera.
+        /// For more info on frustum height, see <see cref="docs.unity3d.com/Manual/FrustumSizeAtDistance.html"/> 
+        /// </summary>
+        /// <param name="state">CameraState for checking if Orthographic or Perspective</param>
+        /// <param name="vcam">vcam, to check its position</param>
+        /// <returns>Frustum height of the camera</returns>
+        private float CalculateHalfFrustumHeight(in CameraState state, in CinemachineVirtualCameraBase vcam)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = Mathf.Abs(state.Lens.OrthographicSize);
+            }
+            else
+            {
+                float distance = m_BoundingShape2D.transform.position.z - vcam.transform.position.z;
```

Comment:
I think you should use actual distance, and not assume that vcam is looking along world z axis.